Various components: Avoid re-renderings from interaction events

danielchalmers opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves dotnet/aspnetcore#1337 or fixes dotnet/aspnetcore#9310 -->

Follows up on dotnet/aspnetcore#8203 by making click and touch events non-rendering for other key controls in order to work around the problem described in dotnet/aspnetcore#8151 and https://github.com/dotnet/maui/issues/31944.

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

Existing unit tests were used

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [ ] I've added relevant tests.

danielchalmers committed:
WIP

danielchalmers commented:
@ScarletKuro Thanks for identifying the fix for MudButton in #8203! Seemed to do the trick as seen below:

https://github.com/MudBlazor/MudBlazor/assets/7112040/ad7456e6-4c5e-4e80-8c79-b80dd631a4af

This PR will expand that effort to other controls. I use MudIconButton and MudMenuItem specifically.

It hasn't been tested at all and pretty sure it doesn't compile, but I'm waiting to see if the original fix has no side effects, as you mentioned, and if you're interested in seeing the effort continued :)

mikes-gh commented:
Thanks for the PR .  For the variable I prefer `_navigating` but it's not critical. @henon ?

henon commented:
@mikes-gh I think you reviewed the commit which Daniel made on his own repo/app ;)

ScarletKuro commented on src/MudBlazor/Components/NavMenu/MudNavLink.razor:
```diff
@@ -8,7 +8,7 @@
 #pragma warning restore CS0618
         { 
-            <NavLink @onclick="@HandleNavigation" 
+            <NavLink @onclick="EventUtil.AsNonRenderingEventHandler<MouseEventArgs>(HandleNavigation)" 
```
Since `HandleNavigation` doesn't use the arguments the `<MouseEventArgs>` needs to be removed

ScarletKuro commented on src/MudBlazor/Components/NavMenu/MudNavLink.razor:
```diff
@@ -24,7 +24,7 @@
         else
         {
-            <div @onclick="OnClickHandler"
+			<div @onclick="EventUtil.AsNonRenderingEventHandler<MouseEventArgs>(OnClickHandler)"
```
Wrong indent

ScarletKuro commented on src/MudBlazor/Components/Menu/MudMenuItem.razor:
```diff
@@ -1,6 +1,6 @@
 @inherits MudComponentBase
 
-<MudListItem @attributes="UserAttributes" @ontouchstart="OnTouchStartHandler" @ontouchmove="OnTouchMoveHandler" @onclick="OnClickHandler" @ontouchend="@OnTouchHandler" Disabled="@Disabled" Class="@Class" Style="@Style" Icon="@Icon" IconColor="@IconColor" IconSize="@IconSize">
+<MudListItem @attributes="UserAttributes" @ontouchstart="OnTouchStartHandler" @ontouchmove="OnTouchMoveHandler" @onclick="EventUtil.AsNonRenderingEventHandler<MouseEventArgs>(OnClickHandler)" @ontouchend="EventUtil.AsNonRenderingEventHandler<MouseEventArgs>(OnTouchHandler)" Disabled="@Disabled" Class="@Class" Style="@Style" Icon="@Icon" IconColor="@IconColor" IconSize="@IconSize">
```
For touch event it should be `TouchEventArgs`.

danielchalmers committed:
Fix extra re-render for more components

danielchalmers committed:
Click docs

danielchalmers committed:
Finishing touches

danielchalmers commented:
@ScarletKuro test is failing with:

```
A total of 1 test files matched the specified pattern.
  Failed AddAndRemoveEventListenerWhenChangingColorPickerView [45 ms]
  Error Message:
   Bunit.Rendering.UnknownEventHandlerIdException : There is no event handler with ID '12' associated with the 'onclick' event in the current render tree.

This can happen, for example, when using `cut.FindAll()`, and calling event trigger methods on the found elements after a re-render of the render tree. The workaround is to use re-issue the `cut.FindAll()` call after each render of a component, this ensures you have the latest version of the render tree and DOM tree available in your test code.
```

which could be from the icon buttons?

https://github.com/MudBlazor/MudBlazor/blob/5b3757a4c881e4b69b4f8eead488e22c1280d905/src/MudBlazor/Components/ColorPicker/MudColorPicker.razor#L12-L21

henon commented:
Yes, we unfortunately have a lot of these stale references in the tests.

ScarletKuro commented:
Yes, this is why I had to do change the [ToggleGroupTests.cs](https://github.com/MudBlazor/MudBlazor/pull/8203/files#diff-40ad308f6a86afedea804bbb5a3da45bb3936c42ff5edeef9e69778142d92b52)
I mentioned this issue here https://github.com/MudBlazor/MudBlazor/discussions/8151#discussioncomment-8501991
and as henon said its incorrectly made unit tests.

danielchalmers committed:
Avoid stale element references in tests

danielchalmers committed:
Avoid stale element references in tests

danielchalmers committed:
revert some changes

danielchalmers committed:
More work on stale references in tests

danielchalmers committed:
More reversions and tests fixed

basically just dialing things back and staying focused on fixing key stale references; less proactive to avoid disrupting the codebase

danielchalmers commented:
@ScarletKuro Think it's about ready if you want to check it out. Hopefully not too big. I am stuck on the last failing test though:

```cs
// MenuTests -> OnClickErrorContentCaughtException
public async Task OnClickErrorContentCaughtException()
{
    var comp = Context.RenderComponent<MenuErrorContenCaughtException>();
    await comp.FindAll("button.mud-button-root")[0].ClickAsync(new MouseEventArgs());
    comp.FindAll("div.mud-popover-open").Count.Should().Be(1);
    comp.FindAll("div.mud-list-item").Count.Should().Be(1);
    await comp.FindAll("div.mud-list-item")[0].ClickAsync(new MouseEventArgs());
    var mudAlert = comp.FindComponent<MudAlert>();
    var text = mudAlert.Find("div.mud-alert-message");
    text.InnerHtml.Should().Be("Oh my! We caught an error and handled it!");
}
```

```razor
@* MenuErrorContenCaughtException *@
@namespace MudBlazor.UnitTests.TestComponents
<ErrorBoundary>
    <ChildContent>
        <MudPopoverProvider></MudPopoverProvider>
        <MudMenu Label="Open Menu">
            <ChildContent>
                <MudMenuItem @onclick="@HandleClickAsync">Throw error</MudMenuItem>
            </ChildContent>
        </MudMenu>
        <br/>
    </ChildContent>
    <ErrorContent>
        <MudAlert Severity="Severity.Error">Oh my! We caught an error and handled it!</MudAlert>
    </ErrorContent>
</ErrorBoundary>

@code {
    public async Task HandleClickAsync()
    {
        await Task.Delay(100);
        throw new Exception("Something went wrong...");
    }
}
```

```
 OnClickErrorContentCaughtException
   Source: MenuTests.cs line 242
   Duration: 111 ms

  Message: 
Bunit.Rendering.ComponentNotFoundException : A component of type MudAlert was not found in the render tree.

  Stack Trace: 
TestRenderer.FindComponent[TComponent](IRenderedFragmentBase parentComponent) line 181
RenderedFragmentExtensions.FindComponent[TComponent](IRenderedFragment renderedFragment) line 58
MenuTests.OnClickErrorContentCaughtException() line 249
GenericAdapter`1.BlockUntilCompleted()
NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaiter)
AsyncToSyncAdapter.Await[TResult](Func`1 invoke)
TestMethodCommand.Execute(TestExecutionContext context)
DelegatingTestCommand.RunTestMethodInThreadAbortSafeZone(TestExecutionContext context, Action action)
```

I don't think it's capturing the exception properly due to the async click handler but I could be wrong. Not sure where to go from here. Adding a `Comp.Render()` in before searching for the Alert will throw the "Something went wrong..." exception.

danielchalmers committed:
update

danielchalmers committed:
start from scratch with unit tests to simplify them

danielchalmers committed:
fix unit tests

mostly just the buttons

MenuTests.OnClickErrorContentCaughtException is the only one remaining

ScarletKuro commented:
I'm slightly confused why are we modifying the examples from Docs, they are not related to the Unit test area.
If it's only for testing purpose then they shouldn't be pushed to remote.
Otherwise I would refrain from bloating the docs, like the alert with button count has no value, same  as the added snackbar to the badge click.

Maybe I'm wrong, @henon can correct me.



> @ScarletKuro @henon Should be ready except for the last failing test:

I will try to look on weekends when I get time, for now I'm busy with other PRs.

danielchalmers committed:
update

danielchalmers committed:
revert some docs

danielchalmers committed:
misc changes and revert list item due to expand toggle not working

makes all changes to SelectTests obsolete and revertable

also fixes the elusive OnClickErrorContentCaughtException

trims down the PR big time. if necessary we can work on list item afterwards but it's better to get the majority through smoothly

danielchalmers committed:
revert overlay

danielchalmers committed:
update

danielchalmers commented:
List item was reverted because the expand toggle wasn't working. Could be looked into further but I don't need it for what I'm doing and it seems to have wide-reaching consequences. Fixes the failing test and simplifies the PR significantly through `SelectTests` not needing the changes made.

Overlay was reverted too bc I'm not comfortable with the consequences of that either considering a `Render` had to be added to the test to make it work. Similar for SwipeArea.

----

@ScarletKuro PR has been trimmed down and all tests are passing 🎉
Two docs added: NavMenuOnClickExample and ChipClickableExample. Will delete if that's preferred!

henon commented:
@danielchalmers can you please post screenshots of the new examples?

henon commented:
OK, I think we can keep these examples.

ScarletKuro committed:
Cosmetics

ScarletKuro committed:
cosmetics 2

ScarletKuro reviewed (approved):
@henon this is ready to merge.

henon commented:
Thanks @danielchalmers
